### PR TITLE
ieee802154: Add support for default flag settings

### PR DESCRIPTION
**Pull request by bergzand**
### Contribution description

This PR adds a number of configuration flags to `ieee802154.h` similar to the current default `PANID` and `CHANNEL` define. This allows for configuring the `AUTOACK`, `ACK_REQ` and `CSMA` flag at compile time. 

Except for Openthread, the current set of supported network stacks are updated to set these flags on interface initialization.

### Issues/PRs references

- None

#### samr21-xpro

Remove [L81-L83](https://github.com/RIOT-OS/RIOT/blob/master/drivers/at86rf2xx/at86rf2xx.c#L81-L83) from the at86rf2xx driver to prevent it from setting its own default. Also keep in mind that there is the issue piggybacked with #9581 where modifying the `AUTOACK` flag influences the `ACK_REQ` flag.

#### Other radios

The other radios have different defaults enabled in their reset functions which should be removed to effectively test this PR.

### Known issues

- For some reason the `AUTOACK` flag is always enabled on the at86rf2xx independent of the settings provided in this PR, I haven't tested yet if the other radios are also affected, but it could be a default setting from the radio hardware itself.

- I don't really like the code duplication between the stack initialization functions but at the moment I don't know a nice place to put a common settings init function

- @jia200x Does OpenThread require any specific flags from the network interface to function, does it make sense to honor the settings provided at compile time as much as possible except for any flags strictly required?

### CODE REVIEW

**Comment by bergzand** [comment]
As a follow up I'd like to extend this to also set the channel and PAN ID this way.

File: sys/net/gnrc/netif/gnrc_netif.c [hunk 1]
```diff
@@ -1240,6 +1263,16 @@ static void _configure_netdev(netdev_t *dev)
...
 #endif
+
+    uint16_t dev_type;
+    res = dev->driver->get(dev, NETOPT_DEVICE_TYPE, &dev_type, sizeof(dev_type));
+    assert(res == sizeof(dev_type));
+
+    switch(dev_type) {
+        case NETDEV_TYPE_IEEE802154:
+            _configure_netdev_802154(dev);
+            break;
```
Comment (bergzand) [inline sys/net/gnrc/netif/gnrc_netif.c]: > Maybe there should be some ifdef for removing the dead code if we are building an application without 802.15.4 (e.g. Ethernet only), to reduce the rom usage

I can put a `#ifdef MODULE_NETDEV_IEEE802154` around the case here and the configuration function above.

File: sys/net/gnrc/netif/gnrc_netif.c [hunk 2]
```diff
@@ -1226,6 +1226,29 @@ static void _init_from_device(gnrc_netif_t *netif)
     _update_l2addr_from_dev(netif);
 }
 
+static void _configure_netdev_802154(netdev_t *dev)
+{
+    netopt_enable_t enable = NETOPT_ENABLE;
```
[the remark this answers is not in the document]
Comment (bergzand) [reply sys/net/gnrc/netif/gnrc_netif.c]: Ack

**Comment by bergzand** [comment]
@gebart fixed! I've also added an ifdef around the config in LwIP. I've used `MODULE_LWIP_SIXLOWPAN` there as it was already used as guards around 802.15.4 specific code

**Comment by miri64** [comment]
I actually already dislike the `ifdef`ing in the current `gnrc_netif.c` and will cause more problems for new device types with similar but not equal options. Maybe this can also be put into the `init` function of the `netdev_data` layer?

**Comment by bergzand** [comment]
> I actually already dislike the ifdefing in the current gnrc_netif.c

Yeah, I agree with this.

> Maybe this can also be put into the init function of the netdev_data layer?

What exactly is the `netdev_data` layer here?

I strongly prefer to keep the initialization of the device settings outside of netdev. When we have the layered netdev approach, it can be dangerous to set netopt settings halfway the stack. Layers on top of the layer enabling settings might have to be aware of settings that get enabled/disabled or have to mask certain settings for lower layers.

For example a software "AUTOACK" layer would have to be aware of enabling or disabling the `NETOPT_AUTOACK` flag.

Maybe we can move this to the `gnrc_netif_ieee802154.c` somehow.

**Comment by miri64** [comment]
Agreed :-/

**Comment by bergzand** [comment]
@miri64 I've moved the flag setters for gnrc to a new (but already supported) `init` function for `gnrc_netif_ieee802154`.

**Comment by miri64** [comment]
I just tested it. My expectation would be that if I set `IEEE802154_DEFAULT_ENABLE_ACK_REQ=0` that `ACK_REQ` would be disabled on start-up. However, the current behavior seems to be "leave it to the device". Is that intended?

**Comment by miri64** [comment]
> However, the current behavior seems to be "leave it to the device". Is that intended?

@bergzand ping?

**Comment by bergzand** [comment]
> I just tested it. My expectation would be that if I set `IEEE802154_DEFAULT_ENABLE_ACK_REQ=0` that `ACK_REQ` would be disabled on start-up. However, the current behavior seems to be "leave it to the device". Is that intended?

It was originally intended. For now it is probably better to enable/disable every setting the way it is defined.

In the long I'd like to remove the mix of flag settings from each device and only enable the flags required/configured by the network stack.

**Comment by miri64** [comment]
> It was originally intended. For now it is probably better to enable/disable every setting the way it is defined.
>
> In the long I'd like to remove the mix of flag settings from each device and only enable the flags required/configured by the network stack.

Ok. That should however be done ASAP after this PR was merged.

**Comment by miri64** [comment]
> Ok. That should however be done ASAP after this PR was merged.

Ping @bergzand?

**Comment by jia200x** [comment]
> @jia200x Does OpenThread require any specific flags from the network interface to function, does it make sense to honor the settings provided at compile time as much as possible except for any flags strictly required?

There is [this function](https://github.com/RIOT-OS/RIOT/blob/master/pkg/openthread/contrib/platform_radio.c#L326). It ask for the radio capabilities. So, up to a certain level we can specify some flags at compile time and report them to the stack with that function.

Concerning CHANNEL, PANID, etc we might need to think how to integrate them at compile time. We have our own init for OpenThread, so I guess it's possible ;)

**Comment by bergzand** [comment]
> > @jia200x Does OpenThread require any specific flags from the network interface to function, does it make sense to honor the settings provided at compile time as much as possible except for any flags strictly required?
> 
> There is [this function](https://github.com/RIOT-OS/RIOT/blob/master/pkg/openthread/contrib/platform_radio.c#L326). It ask for the radio capabilities. So, up to a certain level we can specify some flags at compile time and report them to the stack with that function.

@jia200x But from the set of {`ACK_REQ`, `AUTOACK`, `CSMA`}, is there anything that should always be enable if possible for Openthread to function?

**Comment by jia200x** [comment]
> @jia200x But from the set of {ACK_REQ, AUTOACK, CSMA}, is there anything that should always be enable if possible for Openthread to function?

I will check this ASAP. At least I know the ACK_REQ and AUTOACK are required with the current implemetation. 

Ping in in the next week if I don't reply :)

File: sys/net/gnrc/netif/gnrc_netif_ieee802154.c [hunk 3]
```diff
@@ -74,6 +76,31 @@ static gnrc_pktsnip_t *_make_netif_hdr(uint8_t *mhr)
     return snip;
 }
 
+static void _init(gnrc_netif_t *netif)
```
Comment (bergzand) [inline sys/net/gnrc/netif/gnrc_netif_ieee802154.c]: This should also be called when a `NETOPT_STATE_RESET` set is done

**Comment by bergzand** [comment]
@jia200x More ping!

**Comment by jia200x** [comment]
@bergzand The current OpenThread port assumes:
- Automatic Ack handling (with CSMA)
- Ack Timeout event (e.g call to NOACK event)
- ACK_REQ is enabled.

However most of these features are also implemented by OpenThread (but not used in the RIOT implementation). So, in theory there's no strict requirement on {ACK_REQ, AUTOACK, CSMA} set to make it work. But ACK_REQ should be enabled.

The thing is, [otPlatRadioGetCaps](https://github.com/RIOT-OS/RIOT/blob/master/pkg/openthread/contrib/platform_radio.c#L367) should be implemented with netdev->driver->get calls (NETOPT_CSMA, NETOPT_AUTOACK, etc). I plan do it now-ish :)

TL;DR: Let's assume for now only AUTOACK and ACK_REQ are required (I guess it also calls NOACK if failed, doesn't?)
